*** gnosis/xml/pickle/api.py~	2007-11-11 21:50:51.000000000 +0000
--- gnosis/xml/pickle/api.py	2007-12-06 12:03:04.000000000 +0000
*************** class xmlpickle_meta_parser(ContentHandl
*** 646,652 ****
          
      def fatalError(self,exception):
          if self.real_parser is None:
!             print "FATAL ERROR %s" % str(exception)
              # re-raise exception			
              raise
          else:		
--- 646,652 ----
          
      def fatalError(self,exception):
          if self.real_parser is None:
!             #print "FATAL ERROR %s" % str(exception)
              # re-raise exception			
              raise
          else:		
