--- Python-2.6.4/Lib/lib-tk/Tkinter.py	2006-07-18 18:46:31.000000000 +0100
+++ Python-2.6.4/Lib/lib-tk/Tkinter.py.new	2008-12-20 15:01:40.000000000 +0000
@@ -321,7 +321,13 @@
 
     def get(self):
         """Return the value of the variable as a bool."""
-        return self._tk.getboolean(self._tk.globalgetvar(self._name))
+        # CB: modified to support None
+        # (note that CheckButton only seems to work properly with a BooleanVar,
+        #  so I can't use a StringVar to support our 1/0/None booleans)
+        try:
+            return self._tk.getboolean(self._tk.globalgetvar(self._name))
+        except:
+            return self._tk.globalgetvar(self._name)
 
 def mainloop(n=0):
     """Run the main loop of Tcl."""
